### DISCUSSION TITLE
Using Suspense alongside Zustand crashes the app

### DISCUSSION THREAD

**Discussion by sbboli** [pmndrs/zustand#1727] [Q&A]
## Summary

Hi Folks, I'm running into a problem when using Zustand parallel with React Suspense in a NextJS app.

### Minimal Example:

```javascript
import { Suspense, useEffect } from 'react';
import { create } from 'zustand';

const useMyStore = create(() => {
  return { count: 420 };
});

export default function SuspenseAndZustand() {
  const store = useMyStore();

  useEffect(() => {
    // only crashes when setting the state here
    useMyStore.setState({ count: 9000 });
  }, []);

  return (
    <div>
      <p>store count: {store.count}</p>
      {/* works as expected when removing the suspense block */}
      <Suspense>Crash.</Suspense>
    </div>
  );
}
```

This causes the app to crash with:

```
Unhandled Runtime Error
Error: This Suspense boundary received an update before it finished hydrating. This caused the boundary to switch to client rendering. The usual way to fix this is to wrap the original update in startTransition.
```

When removing the (unrelated) suspense block, everything works as expected.

**Comment by dai-shi** [comment]
Is this behavior specific to nextjs or something,
or can it be reproduced with https://codesandbox.io/s/react-new ?

**Reply by sbboli** [reply]
The error can NOT be reproduced with react only.

Here is the failing next version though: https://codesandbox.io/s/holy-shape-br1uqs?file=/pages/index.js

If anyone had a workaround for this, I'd highly appreciated it.

**Comment by chin2km** [comment]
Can still reproduce this issue :( 
using the `React.startTransition` to do the update to store doesn't help either!

**Comment by dai-shi** [comment]
> https://codesandbox.io/p/devbox/holy-shape-br1uqs?file=%2Fpages%2F_app.js

It doesn't seem using Context. I'm not sure if it's the root cause. cc @dbritto-dev

**Reply by dbritto-dev** [reply]
@dai-shi `useEffect` is only available on client components, they should use 'use client' directive

**Reply by dbritto-dev** [reply]
@sbboli would you mind adding 'use client' directive to your component?

**Reply by dai-shi** [reply]
They aren't using RSC, are they?

**Reply by dbritto-dev** [reply]
As far I see, they are. They are using next with pages router

**Reply by dai-shi** [reply]
RSC is only effective with app router, isn't it?

**Reply by dbritto-dev** [reply]
@dai-shi you are right, I believe is SSR

**Reply by dbritto-dev** [reply]
@sbboli @dai-shi I was able to replicate the issue without `zustand` and `useState` , so is not something related to `zustand`

**Reply by dbritto-dev** [reply]
@dai-shi `startTransition` and `useState` works, but `startTransition` and `zustand store` does not work

**Comment by dbritto-dev** [comment]
@sbboli @chin2km would you mind following our nextjs guide? we are open to any feedback

**Comment by dai-shi** [comment]
> @dai-shi `startTransition` and `useState` works, but `startTransition` and `zustand store` does not work

@dbritto-dev By "does not work", do you mean something like forcing sync render? In other words, does it work with `use-zustand`?

**Reply by dbritto-dev** [reply]
@dai-shi I mean that you get the error related to suspense, something interesting is that when you use `useTransition` instead of `startTransition` you get the same error on both, `useState` and `zustand store`

**Reply by dai-shi** [reply]
Thanks for clarification. ~So, it's SSR related.~

**Reply by dbritto-dev** [reply]
@dai-shi yep, `use-zustand` works well, might be something related to `useSyncExternalStore` package

**Reply by dai-shi** [reply]
Oh, I see. Then, it's somewhat a known behavior.

**Reply by dbritto-dev** [reply]
btw, here you go -> https://codesandbox.io/p/devbox/cranky-tom-mjdcq8
